qutip/tests/test_operators.py

```diff
@@ -79,51 +88,38 @@ def test_jmat_32():
                          [0.0 + 0.j, 0.0 + 0.j, -0.5 + 0.j, 0.0 + 0.j],
                          [0.0 + 0.j, 0.0 + 0.j, 0.0 + 0.j, -1.5 + 0.j]])
 
-    assert_equal(np.allclose(spin32[0].full(), paulix32), True)
-    assert_equal(np.allclose(spin32[1].full(), pauliy32), True)
-    assert_equal(np.allclose(spin32[2].full(), pauliz32), True)
-
-
-def test_jmat_42():
-    "Spin 2 operators"
-    spin42 = jmat(4 / 2., '+')
-    assert_equal(spin42.dims == [[5], [5]], True)
-
-
-def test_jmat_52():
-    "Spin 5/2 operators"
-    spin52 = jmat(5 / 2., '+')
-    assert_equal(spin52.shape == (6, 6), True)
-
-
-def test_destroy():
-    "Destruction operator"
-    b4 = basis(5, 4)
-    d5 = destroy(5)
-    test1 = d5 * b4
-    assert_equal(np.allclose(test1.full(), 2.0 * basis(5, 3).full()), True)
-    d3 = destroy(3)
-    matrix3 = np.array(
-        [[0.00000000 + 0.j, 1.00000000 + 0.j, 0.00000000 + 0.j],
-         [0.00000000 + 0.j, 0.00000000 + 0.j, 1.41421356 + 0.j],
-         [0.00000000 + 0.j, 0.00000000 + 0.j, 0.00000000 + 0.j]])
-
-    assert_equal(np.allclose(matrix3, d3.full()), True)
-
-
-def test_create():
-    "Creation operator"
-    b3 = basis(5, 3)
-    c5 = create(5)
-    test1 = c5 * b3
-    assert_equal(np.allclose(test1.full(), 2.0 * basis(5, 4).full()), True)
-    c3 = create(3)
-    matrix3 = np.array(
-        [[0.00000000 + 0.j, 0.00000000 + 0.j, 0.00000000 + 0.j],
-         [1.00000000 + 0.j, 0.00000000 + 0.j, 0.00000000 + 0.j],
-         [0.00000000 + 0.j, 1.41421356 + 0.j, 0.00000000 + 0.j]])
-
-    assert_equal(np.allclose(matrix3, c3.full()), True)
+    np.testing.assert_allclose(spin32[0].full(), paulix32)
+    np.testing.assert_allclose(spin32[1].full(), pauliy32)
+    np.testing.assert_allclose(spin32[2].full(), pauliz32)
+
+
+@pytest.mark.parametrize(['spin', 'N'], [
+    pytest.param(3/2., 4, id="1.5"),
+    pytest.param(5/2., 6, id="2.5"),
+    pytest.param(3.0, 7, id="3.0"),
+])
+def test_jmat_dims(spin, N):
+    spin_mat = qutip.jmat(spin, '+')
+    assert spin_mat.dims == [[N], [N]]
+    assert spin_mat.shape == (N, N)
+
+
+@pytest.mark.parametrize(['oper_func', 'diag', 'offset', 'args'], [
+    pytest.param(qutip.destroy, np.arange(1, N)**0.5, 1, (), id="destroy"),
+    pytest.param(qutip.destroy, np.arange(6, N+5)**0.5, 1, (5,),
+                 id="destroy_offset"),
+    pytest.param(qutip.create, np.arange(1, N)**0.5, -1, (), id="create"),
+    pytest.param(qutip.create, np.arange(6, N+5)**0.5, -1, (5,),
+                 id="create_offset"),
+    pytest.param(qutip.num, np.arange(N), 0, (), id="num"),
+    pytest.param(qutip.num, np.arange(5, N+5), 0, (5,), id="num_offset"),
+    pytest.param(qutip.charge, np.arange(-N, N+1), 0, (), id="charge"),
+    pytest.param(qutip.charge, np.arange(2, N+1)/3, 0, (2, 1/3),
+                 id="charge_args"),
+])
+def test_diagonal_oper(oper_func, diag, offset, args):
```

Comment:
Can we call it `test_diagonal_operators` (and similar for all tests) - the names appear in the log, so they're basically the top-level test documentation.  We never have to type them, so it's probably better to spell everything out in full.

Sidenote: some languages have syntax to let you define a "literal" identifier, which can be an arbitrary string including spaces, or a single keyword!  I wish we had them in Python - it's really nice to have (e.g.)
```fsharp
let ``test diagonal operators are created correctly`` func diag offset args =
  let operator = func diag offset args
  assert oper == (np.diag diag offset |> qutip.Qobj)
```
(sidesidenote: apparently GitHub's F# highlighting is a little naff)

Reply:
We could get something similar for tests using https://pypi.org/project/pytest-testdox/ if we wanted.